README.md
[![Version][npm-version-img]][npm] [![Downloads][npm-downloads-img]][npm]

[npm]: https://www.npmjs.com/package/vue3-spatialnavigation
[npm-downloads-img]: https://img.shields.io/npm/dm/vue3-spatialnavigation.svg
[npm-version-img]: https://img.shields.io/npm/v/vue3-spatialnavigation.svg

# Introduction
This is a Vue3 Javascript-only adaptation of the Vue2 Spatial Navigation directive from https://github.com/twcapps/vue-spatialnavigation

## Installation

```console
npm install vue3-spatialnavigation --save
```

## Spatial navigation
Spatial navigation is keyboard navigation of a html page most often used on devices where mouse navigation is not available or not preferred. An important rule is that only 1 element can have the active 'focus' state at the same time.

When a user navigates the page with the left, right, up, down keybaord keys a decision should be made which element to focus next. This directive will improve the Vue native keyboard binding to provide a flexible spatial navigation implementation.

Inspiration for this plugin is taken from the [CSS3 draft keyboard navigation](https://drafts.csswg.org/css-ui/#keyboard), a limitation is that CSS3 can only point the focus position based on (DOM) element id pointers, while this directive supports both element id pointers and custom (component) method binding.


## Typical usage

To make an element 'focusable' just add the focus directive
```javascript
import { createApp } from 'vue'
import VueSpatialNavigation from "vue3-spatialnavigation";
const app = createApp();
app.use(VueSpatialNavigation);
```

```html
<Component id="button1" v-spat></Component>
<Component id="button2" v-spat></Component>
```

### CSS3 Element binding
Use data properties with a domId string reference to determine the focus action during navigation (this behaves similar to the CSS3 draft)
```html
<Component id="button1" v-spat data-down="button2"></Component>
<Component id="button2" v-spat data-up="button1"></Component>
```

### JS Component binding
Use vue bind directive with a component function reference to determine the focus action during navigation.
```html
<Component id="button1" v-spat @down="componentFunctionDown"></Component>
<Component id="button2" v-spat @up="componentFunctionUp" ></Component>
```
```javascript
import { createApp } from 'vue'
import VueSpatialNavigation from "vue3-spatialnavigation";
const app = createApp();
app.use(VueSpatialNavigation, {
  methods: {
    componentFunctionDown: () => {
      let el = navigationService.getFocusElementById("button2");
      if (el) el.focus()
    },
    componentFunctionUp: () =>  {
      let el = navigationService.getFocusElementById("button1");
      if (el) el.focus()
    }
  }
})
```

### Automatic focus handling
Use the special constant value 'AUTOFOCUS' to let the directive try to find the next focusable sibbling automatically. This only works if the next/prev focusable element is a sibling within the parent container!
```html
<Component id="button1" v-spat data-down="AUTOFOCUS"></Component>
<Component id="button2" v-spat data-up="AUTOFOCUS"></Component>
```

### Determine the first or default focus element
In case the page is just rendered and or focus is reset the directive should know which element to restore focus to. The directive will keep a reference to the last focussed element, but if that is not available it will restore focus to the 'default' focus element on the page.
```html
<Component id="button1" v-spat data-default data-down="button2"></Component>
<Component id="button2" v-spat data-up="button1"></Component>
```

### Enter action
When a user actions an 'enter' keypress it will automatically be converted into a click action on the element that has active focus. So you can use the regular vue @click bindings.
Note that by default vue only supports @click on html elements and you have to normally use @click.native on a component, this directive will automatically capture the native event though and convert it into a vue event so you can use @click for both native keyboard and mouse clicks.
```html
<Component id="button1" v-spat @down="componentFunctionDown"></Component>
<Component id="button2" v-spat @up="componentFunctionUp" @click="button2clicked"></Component>
```
```javascript
import { createApp } from 'vue'
import VueSpatialNavigation from "vue3-spatialnavigation";
const app = createApp();
app.use(VueSpatialNavigation, {
  methods: {
    componentFunctionDown: () => {
      let el = navigationService.getFocusElementById("button2");
      if (el) el.focus()
    },
    componentFunctionUp: () =>  {
      let el = navigationService.getFocusElementById("button1");
      if (el) el.focus()
    },
    button2clicked: () => {
      console.log("button2clicked");
    }
  }
})
```

## Configuration
The application uses a default keybinding for the d-pad and enter keys, during the directive initialisation this can optionally be overriden with the options object:

```javascript
import { createApp } from 'vue'
import VueSpatialNavigation from "vue3-spatialnavigation";
const app = createApp();
app.use(VueSpatialNavigation, {
  keyCodes: {
    "up": 38,
    "down": 40,
    "left": 37,
    "right": 39,
    "enter": 13
  }
})
```

It is also possible to override the NavigationService class to create a fully custom key binding.

```javascript
import { createApp } from 'vue'
import VueSpatialNavigation from "vue3-spatialnavigation";
import { NavigationService } from "vue3-spatialnavigation/navigation";
// custom navigation service 
class CustomNavigationService extends NavigationService {

}
const app = createApp();
app.use(VueSpatialNavigation, {
  navigationService: CustomNavigationService
})
```


## Styling
The directive will automatically assigned and remove the class "focus" on the html element to distingish an element in focus vs an element not in focus. The css "focus" class handling is fully handled by the directive and should not be changed/set manually.

```html
<style>
.focus { background-color: yellow;}
</style>
<Component id="button1" v-spat data-default data-down="AUTOFOCUS"></Component>
<Component id="button2" v-spat data-up="AUTOFOCUS"></Component>
```

## NavigationService
In addition to the directive this plugin also exposes a NavigationService object that allows control of the focusable elements from JS code. The navigation service keeps track of all focusable elements on the page and can be used to query focusable elements/

It provides the below properties:
```javascript
blockAllSpatialNavigation   // enabled / disable a navigation block (e.g. during loading)
```

It provides the below methods:
```javascript
NavigationService.getFocusElementInFocus()         // get FocusElement currently in focus
NavigationService.getFocusElementById(id)          // get FocusElement by DOM id
NavigationService.getFocusElementIsDefault()       // get FocusElement by default property
NavigationService.blurAllFocusElements()           // blur all FocusElements
```

## FocusElement
FocusElement is a class initialised for each dom element where the directive is applied, you can aquire the FocusElement by quering the NavigationService.

It provides the below properties:
```javascript
  FocusElement.id        // dom id reference
  FocusElement.isFocus   // is element currently focussed
  FocusElement.isDefault // is element a default focusable element
```

And the below properties:
```javascript
  FocusElement.focus()  // manually focus the current element
  FocusElement.blur()   // manually blur the current element
  FocusElement.left()   // manually trigger left action the current element
  FocusElement.right()  // manually trigger right action the current element
  FocusElement.up()     // manually trigger up action the current element
  FocusElement.down()   // manually trigger down action the current element
  FocusElement.enter()  // manually trigger enter action the current element
  ```


CHANGELOG.md
# Changelog

## [0.0.9]
### Changes
* Only manually trigger focus listener when dom event cannot be triggered.

## [0.0.8]
### Changes
* Prevent automatic scroll on DOM focus.

## [0.0.7]
### Fixes
* Fixed focusable element being searched for function action.

## [0.0.6]
### Fixes
* Fix keydown handlers still executing when navigation is blocked.

## [0.0.5]
### Additions
* Exposed DOM event to all focus and navigation event handlers.

## [0.0.4]
### Additions
* Added function check for all event handlers.

### Fixes
* Fixed duplicate @click event being triggered on mouseclick.
* Fixed DOM event not being sent to @click handler.
* Fixed some typos in code.

## [0.0.3]
### Fixes
* Fixed deregistering of focusable elements.

## [0.0.2]
### Fixes
* Fixed multiple enter events sometimes triggering on hitting enter.

## [0.0.1]
### Additions
* Initial release.


navigation.js
export const NavigationServiceDirection = {
    "Up": "up",
    "Down": "down",
    "Left": "left",
    "Right": "right",
    "Enter": "enter"
}

export class NavigationService {
    keyCodes = {};
    focusAbleElements = [];
    lastElementIdInFocus = "";
    blockAllSpatialNavigation = false;

    constructor(keys) {
        // bind keyCodes object from Vue config
        for (let keyName in NavigationServiceDirection) {
            let keyCode = keys[NavigationServiceDirection[keyName]];
            if (keyCode) {
                if (keyCode instanceof Array) {
                    for (let k of keyCode) {
                        this.keyCodes[k] = keyName;
                    }
                } else {
                    this.keyCodes[keyCode] = keyName;
                }
            }
        }

        this.setupKeyBoardEvents();
        this.setupMouseEvents();
    }

    setupKeyBoardEvents() {
        document.addEventListener("keydown", ((e) => {
            // find key code
            const keyCode = e.keyCode ? e.keyCode : e.charCode ? e.charCode : e.which;

            // find key name
            let keyName = this.keyCodes[keyCode];

            // no matching key found
            if (!keyName) return false;

            // spatial navigation is blocked
            if (this.blockAllSpatialNavigation) {
                e.preventDefault();
                e.stopPropagation();
                return false;
            }

            // action spatial navigation
            if (keyName in NavigationServiceDirection) this.spatialNavigationAction(keyName, e);

            e.preventDefault();
        }));
    }

    setupMouseEvents() {
        // enable mouseover event
        document.addEventListener("mouseover", (e) => {
            if (this.blockAllSpatialNavigation) {
                e.preventDefault();
                e.stopPropagation();
                return false;
            }

            let el = this.findFocusable(e.target);
            if (el) el.focus();
        });

        // enable mouseout event
        document.addEventListener("mouseout", (e) => {
            if (this.blockAllSpatialNavigation) {
                e.preventDefault();
                e.stopPropagation();
                return false;
            }

            let el = this.findFocusable(e.target);
            if (el) el.blur();
        });
    }

    // try to find focusable element on mouse hover or click
    findFocusable(target) {
        // inside loop search for focusable element
        // we need this if the focusable element has children inside
        // so e.target can point to child or grandchild of focusable element
        while (target) {
            if (target.id) {
                let focusEl = this.getFocusElementById(target.id);
                if (focusEl) return focusEl;
            }
            if (!target.parentNode) return undefined;
            target = target.parentNode;
        }
        return undefined;
    }

    // action a new spatial navigation action
    spatialNavigationAction(action, e) {
        let el = this.getFocusElementInFocus();

        let keyValue = NavigationServiceDirection[action];

        // initiate focus action if we have active element
        if (el) {
            switch (keyValue) {
                case NavigationServiceDirection.Up:
                    el.up(e);
                    break;
                case NavigationServiceDirection.Down:
                    el.down(e);
                    break;
                case NavigationServiceDirection.Left:
                    el.left(e);
                    break;
                case NavigationServiceDirection.Right:
                    el.right(e);
                    break;
                case NavigationServiceDirection.Enter:
                    el.enter(e);
                    break;
            }

            // if there is no active element, try to find last element in focus
        } else if (this.getFocusElementById(this.lastElementIdInFocus)) {
            const el = this.getFocusElementById(this.lastElementIdInFocus);
            if (el) el.focus();

            // as a last resort, try to find a default focus element to 'reset' focus
        } else {
            const el = this.getFocusElementIsDefault();
            if (el) el.focus();
        }
    }

    // bind focusable component
    registerFocusElement(focusElement) {
        this.focusAbleElements.push(focusElement);
        // set initial focus if there is no active focus and current element is default
        if (focusElement.isDefault && !this.getFocusElementInFocus()) {
            focusElement.focus();
        }
    }

    // unbind focusable component
    deRegisterFocusElement(focusElement) {
        let index = this.focusAbleElements.indexOf(focusElement);
        if (index > -1) {
            let el = this.focusAbleElements.splice(index, 1);
            if (el.length > 0) el[0].destroy();
        }
    }

    // get current component in focus
    getFocusElementInFocus() {
        for (const el of this.focusAbleElements) {
            if (el.isFocus) return el;
        }
    }

    // find focusable component by id
    getFocusElementById(id) {
        for (let el of this.focusAbleElements) {
            if (el.id === id) return el;
        }
    }

    // find component that should be focussed by default
    getFocusElementIsDefault() {
        for (const el of this.focusAbleElements) {
            if (el.isDefault) return el;
        }
    }

    // blurr all focusable components
    blurAllFocusElements() {
        for (const el of this.focusAbleElements) {
            if (el.isFocus) el.blur();
        }
    }
}


spatialnavigation.directive.js
import {NavigationService} from "./navigation.js";
import { isFunction } from "./utils";

// export navigation service
export let navigationService;

// export focus element
export class FocusElement {
    static AutoFocus = "AUTOFOCUS";

    // private properties
    _$el = null;
    _left = undefined;
    _right = undefined;
    _up = undefined;
    _down = undefined;
    _listeners = {
        focus: false,
        blur: false,
        left: false,
        right: false,
        up: false,
        down: false,
        click: false
    };

    // directive identifier (matches related DOM id)
    id = "";
    // is element 'focussed'
    isFocus = false;
    // is element 'selected'
    isSelect = false;
    // should element be 'focussed' by default on rendering
    isDefault = false;

    // directive initialisation
    constructor(vnode) {
        if (vnode && vnode.props) {
            let el = vnode.el;
            const props = vnode.props;

            // enforce a dom id on all focusable elements, if it does not exist generate an id
            if (!el.id) {
                el.id = "focus-el-" + Math.random().toString(36).replace(/[^a-z]+/g, "").substr(0, 10);
            }

            // cache dom properties in directive
            this.id = el.id;
            this.props = props;

            this.isDefault = (props['data-default'] === "" || props['data-default'] === "true");
            this._left = (props['data-left'] || props['onLeft'] || "");
            this._right = (props['data-right'] || props['onRight'] || "");
            this._up = (props['data-up'] || props['onUp'] ||"");
            this._down = (props['data-down'] || props['onDown'] ||"");

            // do not cache the listener logic to prevent memory leaks
            // instead cache the existence of a specific listener in the directive
            this._listeners = {
                focus: !!(props['data-focus'] || props['onFocus']),
                blur: !!(props['data-blur'] || props['onBlur']),
                left: !!(props['data-left'] || props['onLeft']),
                right: !!(props['data-right'] || props['onRight']),
                up: !!(props['data-up'] || props['onUp']),
                down: !!(props['data-down'] || props['onDown']),
                click: !!(props['onClick'])
            };
        }
    }

    // cleanup when directive is destroyed
    destroy() {
        this.id = null;
        this.props = null;
        this.isDefault = false;
        this.isFocus = false;
        this.isSelect = false;
        this._$el = undefined;
        this._left = undefined;
        this._right = undefined;
        this._up = undefined;
        this._down = undefined;
        this._listeners = {
            focus: false,
            blur: false,
            left: false,
            right: false,
            up: false,
            down: false,
            click: false
        };
    }

    // get dom reference of directive
    get $el() {
        if (this._$el) return this._$el;
        return this._$el = document.getElementById(this.id);
    }

    //// focus handling
    // set focus to element
    focus() {
        // blur other element, can only be 1 element in focus
        navigationService.blurAllFocusElements();
        // store focus action in navigation service so we can restore it if needed
        navigationService.lastElementIdInFocus = this.id;
        // set the current element in focus
        this.isFocus = true;
        if (this.$el) {
            this.$el.className += " focus";


            if (this.$el.tabIndex !== -1 || this.$el.nodeName === "INPUT" || this.$el.nodeName === "TEXTAREA") {

                // set 'native' browser focus on input elements and focusable elements.
                this.$el.focus({ preventScroll: true });
            } else {

                // If regular focus event is not triggereable, manually call the focus listener.
                if (this._listeners.focus && isFunction(this.props.onFocus)) {
                    try {
                        this.props.onFocus();
                    } catch (e) {
                    }
                }
            }
        }
    }

    // remove focus from element
    blur() {
        this.isFocus = false;
        if (this.$el) {
            this.$el.className = this.$el.className.replace(/\s?\bfocus\b/, "");
            if (this._listeners.blur) {
                try {
                    this.props.onBlur();
                } catch (e) {
                }
            }
        }
        // if (this.$el && (this.$el.nodeName === "INPUT" || this.$el.nodeName === "TEXTAREA")) this.$el.blur();
    }

    //// select handling
    // set element as selected
    select() {
        this.isSelect = true;
        if (this.$el) this.$el.className += " select";
    }

    // remove selected state from element
    deSelect() {
        this.isSelect = false;
        if (this.$el) this.$el.className.replace(/\bselect\b/, "");
    }

    //// spatial navigation
    // move focus to the element/action configured as 'left' from this element
    left(event) {
        // check if we should automatically find next focusable element
        if (this._left === FocusElement.AutoFocus) {
            this.defaultFocusPrevious();
            // check if next focusable element should be set based on a DOM id pointer
        } else if (this._left && !isFunction(this._left)) {
            this.doFocusElement(this._left);
        }
        // check if a event method is binded to the component
        if (this._listeners.left && isFunction(this.props.onLeft)) {
            try {
                this.props.onLeft(event);
            } catch (e) {
            }
        }
    }

    // move focus to the element/action configured as 'right' from this element
    right(event) {
        // check if we should automatically find next focusable element
        if (this._right === FocusElement.AutoFocus) {
            this.defaultFocusNext();
            // check if next focusable element should be set based on a DOM id pointer
        } else if (this._right && !isFunction(this._right)) {
            this.doFocusElement(this._right);
        }
        // check if a event method is binded to the component
        if (this._listeners.right && isFunction(this.props.onRight)) {
            try {
                this.props.onRight(event);
            } catch (e) {
            }
        }
    }

    // move focus to the element/action configured as 'up' from this element
    up(event) {
        // check if we should automatically find next focusable element
        if (this._up === FocusElement.AutoFocus) {
            this.defaultFocusPrevious();
            // check if next focusable element should be set based on a DOM id pointer
        } else if (this._up && !isFunction(this._up)) {
            this.doFocusElement(this._up);
        }
        // check if a event method is binded to the component
        if (this._listeners.up && isFunction(this.props.onUp)) {
            try {
                this.props.onUp(event);
            } catch (e) {
            }
        }
    }

    // move focus to the element/action configured as 'down' from this element
    down(event) {
        // check if we should automatically find next focusable element
        if (this._down === FocusElement.AutoFocus) {
            this.defaultFocusNext();
            // check if next focusable element should be set based on a DOM id pointer
        } else if (this._down && !isFunction(this._down)) {
            this.doFocusElement(this._down);
        }
        // check if an event method is bound to the component
        if (this._listeners.down && isFunction(this.props.onDown)) {
            try {
                this.props.onDown(event);
            } catch (e) {
            }
        }
    }

    enter(event) {
        if (this._listeners.click && isFunction(this.props.onClick)) {
            try {
                this.props.onClick(event);
            } catch (e) {
            }
        }
    }

    defaultFocusNext() {
        if (this.$el) {
            // check if we can find a sibling element
            const next = this.$el.nextElementSibling;
            // check if element exist and has id
            if (next && next.id) {
                // set focus to component
                this.doFocusElement(next.id);
            }
        }
    }

    defaultFocusPrevious() {
        if (this.$el) {
            // check if we can find a sibling element
            const previous = this.$el.previousElementSibling;
            // check if element exist and has current directive selector
            if (previous && previous.id) {
                // set focus to component
                this.doFocusElement(previous.id);
            }
        }
    }

    doFocusElement(id) {
        let el = navigationService.getFocusElementById(id);
        if (el) el.focus();
    }
}

// Vue plugin
export default {
    install: function (app, options) {
        if (!options) options = {};
        // initialise navigation service
        if (!options.keyCodes) {
            options.keyCodes = {
                "up": 38,
                "down": 40,
                "left": 37,
                "right": 39,
                "enter": 13
            };
        }
        navigationService = (options.navigationService) ? new options.navigationService(options.keyCodes) : new NavigationService(options.keyCodes);

        // Expose navigation service globally.
        app.config.globalProperties.$nav = navigationService;

        app.directive("spat", {
            // directive lifecycle
            mounted: (el, binding, vnode) => {
                let focusElement = new FocusElement(vnode);
                navigationService.registerFocusElement(focusElement);

                // set this element in focus if no element has focus and this is marked default
                if (focusElement.isDefault && !navigationService.getFocusElementInFocus()) {
                    focusElement.focus();
                }
            },
            beforeUnmount: (el, binding, vnode) => {
                if (el.id) {
                    let focusElement = navigationService.getFocusElementById(el.id);
                    if (focusElement) navigationService.deRegisterFocusElement(focusElement);
                }
            }
        });
    }
};
